toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -217,6 +218,85 @@ func SetupLoopbackDevice(diskFilePath string) (devicePath string, err error) {
 	return
 }
 
+// BlockOnDiskIO waits until all outstanding operations against a disk complete.
+func BlockOnDiskIO(diskDevPath string) (err error) {
+	const (
+		// Indices for values in /proc/diskstats
+		majIdx            = 0
+		minIdx            = 1
+		outstandingOpsIdx = 11
+	)
+	var blockDevices blockDevicesOutput
+
+	logger.Log.Infof("Flushing all IO to disk for %s", diskDevPath)
+	shell.Execute("sync")
+
+	rawDiskOutput, stderr, err := shell.Execute("lsblk", "--nodeps", "--json", "--output", "NAME,MAJ:MIN", diskDevPath)
+	if err != nil {
+		logger.Log.Warn(stderr)
+		return
+	}
+
+	bytes := []byte(rawDiskOutput)
+	err = json.Unmarshal(bytes, &blockDevices)
+	if err != nil {
+		return
+	}
+
+	if len(blockDevices.Devices) != 1 {
+		return fmt.Errorf("couldn't find disk IDs for %s (%s), expecting only one result", diskDevPath, rawDiskOutput)
+	}
+	// MAJ:MIN is returned in the form "1:2"
+	diskIDs := strings.Split(blockDevices.Devices[0].MajMin, ":")
+	if len(diskIDs) != 2 {
+		return fmt.Errorf("couldn't find disk IDs for %s (%s), couldn't parse MAJ:MIN", diskDevPath, rawDiskOutput)
+	}
+	maj := diskIDs[0]
+	min := diskIDs[1]
+
+	logger.Log.Tracef("Searching /proc/diskstats for %s (%s:%s)", blockDevices.Devices[0].Name, maj, min)
+	for {
+		var (
+			foundEntry     = false
+			outstandingOps = ""
+		)
+
+		// Find the entry with Major#, Minor#, ..., IOs which matches our disk
+		onStdout := func(args ...interface{}) {
+
+			// Bail early if we already found the entry
+			if foundEntry {
+				return
+			}
+
+			line := args[0].(string)
+			deviceStatsFields := strings.Fields(line)
+			if maj == deviceStatsFields[majIdx] && min == deviceStatsFields[minIdx] {
+				outstandingOps = deviceStatsFields[outstandingOpsIdx]
+				foundEntry = true
+			}
+		}
+
+		err = shell.ExecuteLiveWithCallback(onStdout, logger.Log.Error, true, "cat", "/proc/diskstats")
+		if err != nil {
+			logger.Log.Error(stderr)
+			return
+		}
+		if !foundEntry {
+			return fmt.Errorf("couldn't find entry for '%s' in /proc/diskstats", diskDevPath)
+		}
+		logger.Log.Debugf("Outstanding operations on '%s': %s", diskDevPath, outstandingOps)
+
+		if outstandingOps == "0" {
+			break
+		} else {
```

Comment:
>else [](start = 4, length = 4)

Nit: not need for the `else` block, since we're going to break out of the loop anyway, if `outstandingOps == "0"`.